src/bug_notifier.cpp

```diff
@@ -0,0 +1,32 @@
+#include <set>
+#include <memory>
+
+#include <simulator/bug_notifier.hpp>
+#include <simulator/bug.hpp>
+#include <simulator/bug_observer.hpp>
+
+namespace simulation 
+{
+  BugNotifier::BugNotifier() {}
+
+  void BugNotifier::add_observer(std::shared_ptr<BugObserver> observer) {
+    _observers.insert(observer);
+  }
+
+  void BugNotifier::remove_observer(std::shared_ptr<BugObserver> observer) {
+    _observers.erase(observer);
+  }
+
+  void BugNotifier::notify_birth(std::shared_ptr<Bug> newborn_bug) {
+    for(const auto observer : _observers) {
+      observer->on_bug_born(newborn_bug);
```

Comment:
Shall we try to fix this or is it okay as its not part of a public API and we're not doing blocking calls in this codebase?

If we decide to refactor/change/update this, I'd appreciate to know what your take on this would be (or at least a hint?). I'm not too proficient in multi-threading code.

Reply:
Multithreading is out of scope for this padawan, so let's leave it as is but keep the warning for future designs in more complex systems.